File: android/app/src/main/java/com/openburnbar/ui/components/LiquidGlass.kt
package com.openburnbar.ui.components

import android.os.Build
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.Stable
import androidx.compose.runtime.compositionLocalOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.BlurEffect
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.graphics.TileMode
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.translate
import androidx.compose.ui.graphics.layer.GraphicsLayer
import androidx.compose.ui.graphics.layer.drawLayer
import androidx.compose.ui.graphics.rememberGraphicsLayer
import androidx.compose.ui.layout.onGloballyPositioned
import androidx.compose.ui.layout.positionInRoot
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.openburnbar.ui.theme.AuroraColors
import com.openburnbar.ui.theme.AuroraGradients
import com.openburnbar.ui.theme.AuroraRadius
import com.openburnbar.ui.theme.AuroraShadowSpec
import com.openburnbar.ui.theme.AuroraShadows

// Liquid Glass adapters — Compose mirror of the iOS shape-level vocabulary in
// `OpenBurnBarMobile/Theme/LiquidGlass.swift`. Keep the two files in lockstep
// when the vocabulary grows.
//
// Vocabulary:
//   • `Modifier.liquidGlassSurface(shape)`        — glass plate for passive
//     surfaces: trays, floating bars, cards, sheet inserts.
//   • `Modifier.liquidGlassInteractive(tint, shape)` — glass for tappable
//     controls: buttons, chips, pills. Pass `tint` only to convey meaning
//     (primary action, destructive), never decoration.
//   • `Modifier.liquidGlassCircleButton(diameter)` — the recurring circular
//     overlay control (close ✕, call controls, FAB discs).
//   • `LiquidGlassGroup { }`                       — shared sampling scope.
//     Mark the content the glass floats over with `Modifier.liquidGlassBackdrop()`
//     and, on API 31+, every glass child inside the group draws a genuinely
//     blurred copy of that backdrop under its plate. Below 31 (and outside a
//     group) the adapters fall back to the flat translucent stack.
//
// Why the blur works here when `auroraGlass` dropped it: `RenderEffect` on a
// composable's own layer blurs the composable's CONTENT, not the backdrop —
// that experiment made cards muddy and was removed. The group instead records
// the backdrop into a `GraphicsLayer` and each glass child re-draws that
// recording — offset into its own frame, blurred, clipped to its shape — which
// is true backdrop sampling, the same direction iOS Liquid Glass refracts.
//
// Brand rule (identical to iOS/macOS): glass is the language of the
// utilitarian shell — trays, toolbars, chips, overlay chrome. The
// membership/Pro foil & holo world (`ui/pro`, the store tier cards) stays
// glassless; there, glass appears only on system chrome such as close buttons.

private const val SURFACE_BASE_ALPHA = 0.48f
private const val INTERACTIVE_TINT_ALPHA = 0.22f
private const val DARK_FILL_DAMPING = 0.35f
private const val SAMPLED_FILL_DAMPING = 0.5f
private const val LIGHT_SHEEN_TOP_ALPHA = 0.18f
private const val DARK_SHEEN_TOP_ALPHA = 0.05f
private const val LIGHT_SHEEN_BOTTOM_ALPHA = 0.05f
private const val DARK_SHEEN_BOTTOM_ALPHA = 0.03f
private const val EDGE_STROKE_WIDTH_DP = 0.75f
private val BackdropBlurRadius = 18.dp

/** Default circular control size, mirroring the iOS adapter's 30pt default. */
private val CircleButtonDiameter = 30.dp

// MARK: - Grouping container (backdrop sampling scope)

/**
 * Links a group's recorded backdrop to the glass children that sample it.
 * `frame` is bumped each time the backdrop re-records so children re-draw.
 */
@Stable
internal class LiquidGlassBackdropState {
    var layer: GraphicsLayer? = null
    var sourcePositionInRoot: Offset by mutableStateOf(Offset.Zero)
    var frame: Int by mutableIntStateOf(0)
}

internal val LocalLiquidGlassBackdrop =
    compositionLocalOf<LiquidGlassBackdropState?> { null }

/**
 * Shared sampling scope for grouped glass elements — the Compose analogue of
 * the iOS `GlassEffectContainer` wrapper. On API 31+ the child marked with
 * [liquidGlassBackdrop] is recorded once and every glass adapter inside the
 * group samples that recording; on older systems content renders unchanged
 * and the adapters use their flat translucent fallback.
 */
@Composable
fun LiquidGlassGroup(
    modifier: Modifier = Modifier,
    content: @Composable BoxScope.() -> Unit,
) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
        val state = remember { LiquidGlassBackdropState() }
        state.layer = rememberGraphicsLayer()
        CompositionLocalProvider(LocalLiquidGlassBackdrop provides state) {
            Box(modifier) { content() }
        }
    } else {
        Box(modifier) { content() }
    }
}

/**
 * Marks the content a [LiquidGlassGroup]'s glass children float over. Apply to
 * the backdrop composable (live canvas, artwork, scroll content) — NOT to the
 * glass elements themselves, or they would sample their own pixels.
 */
@Composable
fun Modifier.liquidGlassBackdrop(): Modifier {
    val state = LocalLiquidGlassBackdrop.current ?: return this
    val layer = state.layer ?: return this
    return this
        .onGloballyPositioned { state.sourcePositionInRoot = it.positionInRoot() }
        .drawWithContent {
            layer.record { this@drawWithContent.drawContent() }
            drawLayer(layer)
            state.frame += 1
        }
}

// MARK: - Shape-level adapters

/**
 * Glass plate for a passive surface (tray, floating bar, card). Inside a
 * [LiquidGlassGroup] on API 31+ the plate sits on a blurred backdrop sample;
 * everywhere else it falls back to the flat translucent Aurora stack.
 *
 * `wash` is the sanctioned escape hatch from the iOS inline-glass pattern: a
 * status or smoke color (alpha included by the caller) riding on the plate for
 * legibility over bright content — e.g. `Color.Black.copy(alpha = 0.3f)` for
 * capsules floating over a live mirror. Never use it for decoration.
 */
@Composable
fun Modifier.liquidGlassSurface(
    shape: Shape = RoundedCornerShape(AuroraRadius.lg.dp),
    wash: Color? = null,
    washBrush: Brush? = null,
    shadow: AuroraShadowSpec = AuroraShadows.none,
    isDark: Boolean = isSystemInDarkTheme(),
): Modifier =
    liquidGlassPlate(
        shape = shape,
        wash = wash,
        washBrush = washBrush,
        shadow = shadow,
        isDark = isDark,
    )

/**
 * Glass for a tappable control. Pass `tint` only to convey meaning (primary
 * action, destructive) — glass chrome stays monochrome by default. Press
 * feedback (scale/alpha) stays caller-side, matching the house pattern.
 */
@Composable
fun Modifier.liquidGlassInteractive(
    tint: Color? = null,
    shape: Shape = RoundedCornerShape(AuroraRadius.lg.dp),
    shadow: AuroraShadowSpec = AuroraShadows.none,
    isDark: Boolean = isSystemInDarkTheme(),
): Modifier =
    liquidGlassPlate(
        shape = shape,
        wash = tint?.copy(alpha = INTERACTIVE_TINT_ALPHA),
        washBrush = null,
        shadow = shadow,
        isDark = isDark,
    )

/**
 * The recurring circular glass control used in toolbars and as floating
 * overlay buttons (close ✕, call controls, FAB discs).
 */
@Composable
fun Modifier.liquidGlassCircleButton(
    diameter: Dp = CircleButtonDiameter,
    tint: Color? = null,
): Modifier = size(diameter).liquidGlassInteractive(tint = tint, shape = CircleShape)

// MARK: - Shared plate implementation

@Composable
private fun Modifier.liquidGlassPlate(
    shape: Shape,
    wash: Color?,
    washBrush: Brush? = null,
    shadow: AuroraShadowSpec,
    isDark: Boolean,
): Modifier {
    val backdrop = LocalLiquidGlassBackdrop.current
    val withShadow =
        if (shadow.elevation > 0.dp) {
            this.shadow(
                elevation = shadow.elevation,
                shape = shape,
                spotColor = Color.Black.copy(alpha = shadow.spotAlpha),
                ambientColor = Color.Black.copy(alpha = shadow.spotAlpha),
            )
        } else {
            this
        }
    return withShadow
        .clip(shape)
        .liquidGlassBackdropSample(backdrop)
        .background(baseFill(isDark = isDark, sampled = backdrop?.layer != null), shape)
        .background(sheenBrush(isDark = isDark), shape)
        .liquidGlassWash(wash = wash, washBrush = washBrush, shape = shape)
        .border(EDGE_STROKE_WIDTH_DP.dp, Brush.linearGradient(AuroraGradients.glassStroke), shape)
}

private fun Modifier.liquidGlassWash(wash: Color?, washBrush: Brush?, shape: Shape): Modifier {
    val withColor = if (wash != null) background(wash, shape) else this
    return if (washBrush != null) withColor.background(washBrush, shape) else withColor
}

/**
 * Base translucent wash. When a backdrop sample is being drawn underneath, the
 * wash thins out so the blur shows through; without one it carries the surface
 * on its own (the pre-31 / out-of-group fallback).
 */
private fun baseFill(isDark: Boolean, sampled: Boolean): Color {
    val requested = if (sampled) SURFACE_BASE_ALPHA * SAMPLED_FILL_DAMPING else SURFACE_BASE_ALPHA
    return if (isDark) {
        AuroraColors.darkSurface.copy(alpha = (requested * DARK_FILL_DAMPING).coerceIn(0f, 1f))
    } else {
        AuroraColors.lightSurface.copy(alpha = requested)
    }
}

private fun sheenBrush(isDark: Boolean): Brush =
    Brush.verticalGradient(
        colors =
        listOf(
            Color.White.copy(alpha = if (isDark) DARK_SHEEN_TOP_ALPHA else LIGHT_SHEEN_TOP_ALPHA),
            Color.Transparent,
            AuroraColors.blaze.copy(
                alpha = if (isDark) DARK_SHEEN_BOTTOM_ALPHA else LIGHT_SHEEN_BOTTOM_ALPHA,
            ),
        ),
    )

// MARK: - Backdrop sampling (API 31+)

@Composable
private fun Modifier.liquidGlassBackdropSample(state: LiquidGlassBackdropState?): Modifier {
    if (state == null || Build.VERSION.SDK_INT < Build.VERSION_CODES.S) return this
    val blurLayer = rememberGraphicsLayer()
    var positionInRoot by remember { mutableStateOf(Offset.Zero) }
    return this
        .onGloballyPositioned { positionInRoot = it.positionInRoot() }
        .drawBehind { drawBlurredBackdrop(state, blurLayer, positionInRoot) }
}

private fun DrawScope.drawBlurredBackdrop(
    state: LiquidGlassBackdropState,
    blurLayer: GraphicsLayer,
    positionInRoot: Offset,
) {
    val source = state.layer ?: return
    // Draw-phase read: re-draws this plate whenever the backdrop re-records.
    if (state.frame < 0) return
    val offset = state.sourcePositionInRoot - positionInRoot
    blurLayer.renderEffect =
        BlurEffect(
            radiusX = BackdropBlurRadius.toPx(),
            radiusY = BackdropBlurRadius.toPx(),
            edgeTreatment = TileMode.Clamp,
        )
    blurLayer.record {
        translate(left = offset.x, top = offset.y) {
            drawLayer(source)
        }
    }
    drawLayer(blurLayer)
}

File: AgentLens/Theme/LiquidGlass.swift
import AppKit
import SwiftUI
import OpenBurnBarUI

// MARK: - Liquid Glass (macOS 26+) adapters
//
// macOS mirror of `OpenBurnBarMobile/Theme/LiquidGlass.swift` — keep the two
// files in lockstep when the vocabulary grows. The app deploys to macOS 14,
// so every glass API is gated on `#available(macOS 26, *)` with a material
// fallback that approximates the look on older systems.
//
// Vocabulary:
//   • `liquidGlassSurface(in:fallback:)`     — glass plate for passive
//     surfaces: trays, floating bars, cards, sheet inserts.
//   • `liquidGlassInteractive(tint:in:fallback:)` — glass that responds to
//     pointer/touch, for clickable controls: buttons, chips, pills.
//   • `liquidGlassCircleButton(diameter:)`   — the recurring circular
//     toolbar/overlay control (close ✕, collapse ⌄, etc.).
//   • `liquidGlassEffect(_:in:)`             — drop-in for SwiftUI's
//     `glassEffect(_:in:)` at macOS-26-only call sites, so one-off glass
//     accents honor the transparency preference too.
//   • `LiquidGlassGroup(spacing:)`           — `GlassEffectContainer` when
//     available (glass cannot sample other glass, so grouped elements must
//     share one container); passes content through untouched on macOS 14–15.
//   • `LiquidGlassTransparency`              — the user's glass transparency
//     preference (Frosted ⟷ System ⟷ Clear); see below.
//
// Brand rule: glass is the language of the utilitarian shell — popover cards,
// HUD pills, toolbars, overlay chrome. The membership/Pro world keeps its
// obsidian-foil identity (`Views/Components/Pro`); there, glass appears only
// in system chrome (close buttons, sheet material), never on the foil cards
// themselves.

// MARK: - Transparency preference

/// User-adjustable glass transparency, shared across iOS and macOS through
/// the same UserDefaults key (the two `Theme/LiquidGlass.swift` files mirror
/// each other — keep in lockstep).
///
/// Semantics of the stored value `t` (Double, clamped to -1…1):
///   • `t == 0` — system default. Glass renders exactly as the OS does, which
///     already honors System Settings → Accessibility → Reduce transparency.
///   • `t > 0`  — clearer. The plate switches to the `.clear` glass variant
///     (more see-through); on macOS 14–15 the fallback material plate fades
///     toward the raw backdrop instead.
///   • `t < 0`  — frostier. A thick-material scrim slides in between the
///     plate and the content, approaching an opaque surface at -1.
///
/// Reduce Transparency always wins over "clearer": when the accessibility
/// flag is on, positive values resolve to 0 so glass never becomes *more*
/// transparent than the system allows. Frostier values still apply — they
/// only ever add opacity, which is the direction the flag asks for.
/// Preference math lives in OpenBurnBarUI.LiquidGlassTransparency.
/// This file keeps the platform view adapters only.

/// Optical dimming behind system glass. Renders nothing at `t == 0`.
///
/// Never a `Material`. Material under `glassEffect` is sampled instead of
/// the live canvas, which collapses Liquid Glass into a blur panel.
@ViewBuilder
private func liquidGlassScrim(for t: Double, in shape: some Shape) -> some View {
    let frost = LiquidGlassTransparency.frostScrimOpacity(t)
    let bridge = LiquidGlassTransparency.clearBridgeScrimOpacity(t)
    if frost > 0 {
        shape.fill(Color.black.opacity(min(LiquidGlassTransparency.maximumUnderGlassScrimOpacity, frost)))
    } else if bridge > 0 {
        shape.fill(Color.black.opacity(bridge))
    }
}

private struct LiquidGlassSurfaceModifier<S: Shape>: ViewModifier {
    let tint: Color?
    let shape: S
    let fallback: Material

    @AppStorage(LiquidGlassTransparency.storageKey) private var rawTransparency: Double = 0
    @AppStorage(LiquidGlassTransparency.contentSurfacesEnabledKey) private var contentSurfacesEnabled: Bool = LiquidGlassTransparency.defaultContentSurfacesEnabled
    @Environment(\.accessibilityReduceTransparency) private var reduceTransparency

    func body(content: Content) -> some View {
        let t = LiquidGlassTransparency.effective(rawTransparency, reduceTransparency: reduceTransparency)
        if #available(macOS 26, *), contentSurfacesEnabled {
            let base: Glass = LiquidGlassTransparency.usesClearGlass(t) ? .clear : .regular
            let glass = tint.map { base.tint($0) } ?? base
            content
                .glassEffect(glass, in: shape)
                .background { liquidGlassScrim(for: t, in: shape) }
        } else {
            content
                .background { liquidGlassScrim(for: t, in: shape) }
                .background { if let tint { shape.fill(tint.opacity(0.22)) } }
                .background(fallback.opacity(LiquidGlassTransparency.fallbackPlateOpacity(t)), in: shape)
        }
    }
}

private struct LiquidGlassInteractiveModifier<S: Shape>: ViewModifier {
    let tint: Color?
    let shape: S
    let fallback: Material

    @AppStorage(LiquidGlassTransparency.storageKey) private var rawTransparency: Double = 0
    @AppStorage(LiquidGlassTransparency.contentSurfacesEnabledKey) private var contentSurfacesEnabled: Bool = LiquidGlassTransparency.defaultContentSurfacesEnabled
    @Environment(\.accessibilityReduceTransparency) private var reduceTransparency

    func body(content: Content) -> some View {
        let t = LiquidGlassTransparency.effective(rawTransparency, reduceTransparency: reduceTransparency)
        if #available(macOS 26, *), contentSurfacesEnabled {
            let base: Glass = LiquidGlassTransparency.usesClearGlass(t) ? .clear : .regular
            let glass = (tint.map { base.tint($0) } ?? base).interactive()
            content
                .glassEffect(glass, in: shape)
                .background { liquidGlassScrim(for: t, in: shape) }
        } else {
            content
                .background { liquidGlassScrim(for: t, in: shape) }
                .background { if let tint { shape.fill(tint.opacity(0.22)) } }
                .background(fallback.opacity(LiquidGlassTransparency.fallbackPlateOpacity(t)), in: shape)
        }
    }
}

extension View {
    /// Glass plate for a passive surface (tray, floating bar, card).
    /// Falls back to the given material on macOS 14–15.
    ///
    /// Pass `tint` to lean the plate toward a theme/brand cast (e.g. the
    /// dashboard sidebar leaning on the active layout's signature colour). Keep
    /// it subtle — the tint refracts the backdrop, it does not paint over it.
    func liquidGlassSurface(
        tint: Color? = nil,
        in shape: some Shape,
        fallback: Material = .ultraThinMaterial
    ) -> some View {
        modifier(LiquidGlassSurfaceModifier(tint: tint, shape: shape, fallback: fallback))
    }

    /// Glass for a clickable control. Pass `tint` only to convey meaning
    /// (primary action, destructive), not decoration — toolbar glass is
    /// monochrome by default in the new design.
    func liquidGlassInteractive(
        tint: Color? = nil,
        in shape: some Shape,
        fallback: Material = .ultraThinMaterial
    ) -> some View {
        modifier(LiquidGlassInteractiveModifier(tint: tint, shape: shape, fallback: fallback))
    }

    /// The recurring circular glass control used in toolbars and as floating
    /// overlay buttons (close ✕, collapse chevron, etc.).
    func liquidGlassCircleButton(diameter: CGFloat = 30) -> some View {
        frame(width: diameter, height: diameter)
            .liquidGlassInteractive(in: .circle)
    }
}

// MARK: - Tuned drop-in for direct `glassEffect` call sites

/// Mirror of SwiftUI's `Glass` fluent configuration, so macOS-26-only call
/// sites keep the familiar spelling while routing through the transparency
/// preference: `.liquidGlassEffect(.regular.tint(accent).interactive(), in: .circle)`.
@available(macOS 26.0, *)
struct LiquidGlassStyle {
    var tintColor: Color?
    var isInteractive: Bool
    var clearAtNeutral: Bool

    static var regular: LiquidGlassStyle { .init(tintColor: nil, isInteractive: false, clearAtNeutral: false) }
    static var clear: LiquidGlassStyle { .init(tintColor: nil, isInteractive: false, clearAtNeutral: true) }

    func tint(_ color: Color?) -> LiquidGlassStyle {
        var style = self
        style.tintColor = color
        return style
    }

    func interactive(_ isEnabled: Bool = true) -> LiquidGlassStyle {
        var style = self
        style.isInteractive = isEnabled
        return style
    }

    /// The system glass this style resolves to at transparency `t`.
    ///
    /// `overMediaRichContent` is injectable so no caller — and no test — resolves glass
    /// through a global `UserDefaults` read. A test that routed through the convenience
    /// was green only on a machine with the kernel backdrop switched on.
    func resolvedGlass(
        at t: Double,
        overMediaRichContent: Bool = LiquidGlassTransparency.isOverMediaRichContent()
    ) -> Glass {
        let shouldUseClear = clearAtNeutral
            ? t >= 0
            : LiquidGlassTransparency.usesClearGlass(t, overMediaRichContent: overMediaRichContent)
        var glass: Glass = shouldUseClear ? .clear : .regular
        if let tintColor { glass = glass.tint(tintColor) }
        if isInteractive { glass = glass.interactive() }
        return glass
    }
}

@available(macOS 26.0, *)
private struct LiquidGlassEffectModifier<S: Shape>: ViewModifier {
    let style: LiquidGlassStyle
    let shape: S

    @AppStorage(LiquidGlassTransparency.storageKey) private var rawTransparency: Double = 0
    @AppStorage(LiquidGlassTransparency.contentSurfacesEnabledKey) private var contentSurfacesEnabled: Bool = LiquidGlassTransparency.defaultContentSurfacesEnabled
    @Environment(\.accessibilityReduceTransparency) private var reduceTransparency

    func body(content: Content) -> some View {
        let t = LiquidGlassTransparency.effective(rawTransparency, reduceTransparency: reduceTransparency)
        if contentSurfacesEnabled {
            content
                .glassEffect(style.resolvedGlass(at: t), in: shape)
                .background { liquidGlassScrim(for: t, in: shape) }
        } else {
            content
                .background(.ultraThinMaterial, in: shape)
        }
    }
}

@available(macOS 26.0, *)
extension View {
    /// Drop-in replacement for `glassEffect(_:in:)` that honors the user's
    /// Liquid Glass transparency preference. Use this instead of calling
    /// `glassEffect` directly inside `#available(macOS 26, *)` branches.
    func liquidGlassEffect(_ style: LiquidGlassStyle = .regular, in shape: some Shape) -> some View {
        modifier(LiquidGlassEffectModifier(style: style, shape: shape))
    }

    /// Shape-less overload mirroring `glassEffect(_:)`.
    ///
    /// Substitutes `ConcentricRectangle`, **not** `Capsule`. The system default is
    /// `DefaultGlassEffectShape()`, which concentrically matches the container it sits
    /// in; a capsule is only correct for pill-shaped controls, so the previous version
    /// silently rounded every shape-less call site into a lozenge and drifted away from
    /// system chrome. `ConcentricRectangle` is the closest public equivalent and is
    /// available on the same OS versions as the glass APIs themselves.
    func liquidGlassEffect(_ style: LiquidGlassStyle = .regular) -> some View {
        modifier(LiquidGlassEffectModifier(style: style, shape: ConcentricRectangle()))
    }
}

// MARK: - Behind-window blend (macOS clarity payoff)

/// Blurred desktop showing through the window — the macOS payoff for the
/// "Clear" side of the transparency preference. Layer it at the very back of
/// a window's backdrop and fade it in with the effective adjustment; at
/// `t == 0` callers skip it entirely, so the default window stays opaque.
struct LiquidGlassWindowBlend: NSViewRepresentable {
    func makeNSView(context: Context) -> NSVisualEffectView {
        Self.makeVisualEffectView()
    }

    static func makeVisualEffectView() -> NSVisualEffectView {
        let view = NSVisualEffectView()
        view.material = .underWindowBackground
        view.blendingMode = .behindWindow
        view.state = .active
        return view
    }

    func updateNSView(_ nsView: NSVisualEffectView, context: Context) {}
}

// MARK: - Grouping container

/// Wraps grouped glass elements in a `GlassEffectContainer` on macOS 26 so
/// they share one sampling region (glass cannot sample other glass); on
/// earlier systems the content renders unchanged. `spacing` should match the
/// actual layout spacing of the grouped elements.
struct LiquidGlassGroup<Content: View>: View {
    var spacing: CGFloat?
    @ViewBuilder let content: () -> Content

    init(spacing: CGFloat? = nil, @ViewBuilder content: @escaping () -> Content) {
        self.spacing = spacing
        self.content = content
    }

    var body: some View {
        if #available(macOS 26, *) {
            if let spacing {
                GlassEffectContainer(spacing: spacing, content: content)
            } else {
                GlassEffectContainer(content: content)
            }
        } else {
            content()
        }
    }
}

// NOTE: `GlassCard` / `GlassButton` (Views/Popover/MenuBarPopoverView.swift)
// remain the variant card system for popover/dashboard content — they layer
// the house sheen + edge gradient and adopt real glass on macOS 26 themselves.
// This file holds only the small shape-level adapters; card-level glass rides
// on those existing styles (their macOS 26 paths route through
// `liquidGlassEffect`, so they honor the transparency preference too).
